File: StriverA2Z/C++/6_LinkedList/Questions/SinglyLinkedList/tortoise_and_hare.cpp
#include<bits/stdc++.h>
#include "Node.h"

using namespace std;

Node *middle(Node *head) {
    Node *slow = head;
    Node *fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
    }
    return slow;
}

int main() {
    vector<int> arr = {1,2,3,4,5,6};
    Node *head = array_to_linkedlist(arr);
    Node *mid = middle(head);
    cout << mid->data;
    return 0;
}

File: StriverA2Z/Java/LinkedLists/Questions/SinglyLinkedList/TortoiseAndHare.java
public class TortoiseAndHare {
    public static void main(String[] args) {
        Node head = LinkedListOperations.arrayToLinkedList(new int[] {1,2,3,4,5,6});
        Node mid = middle(head);
        System.out.println(mid.data);
    }

    static Node middle(Node head) {
        Node fast = head;
        Node slow = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }
}
